File: MatrixFormulation/R/UsingSparseMatrix.R
#Fast loading of Data
require(data.table)

#for dealing with Sparse Matrices
require(Matrix)

#probability for visiting a link through edge
beta = 0.8

#load directed edges from the file
rawdata <- fread("../Graph/web-Google.txt",col.names=c("from","to"))

#Number of unique Nodes in the data
nodes <- sort(unique(c(rawdata$from,rawdata$to)))

#Replacing unique nodes with numbers
nodeNumbers <- c(1:length(nodes))
nodeLookup <- data.frame(nodes,nodeNumbers)

#Merging nodeLookup to find replacement of from nodes
fromNodesWithReplacement <- merge(rawdata,nodeLookup,by.x="from",by.y="nodes")

#column names of merged rawdata with from nodes
colnames(fromNodesWithReplacement) <- c("from","to","fromReplacement")

#Merging fromNodesWithReplacement to find replacement of to nodes
allNodesWithReplacement <- merge(fromNodesWithReplacement,nodeLookup,by.x="to",by.y="nodes")

#column names of above merged data with to nodes
colnames(allNodesWithReplacement) <- c("to","from","fromReplacement","toReplacement")

#Calculating number of outlinks for each node
freqOfEachNode <- as.data.frame(table(allNodesWithReplacement$fromReplacement))
freqOfEachNode$Var1 <- as.numeric(as.character(freqOfEachNode$Var1))
reqData <- subset(allNodesWithReplacement, select=c("fromReplacement", "toReplacement"))
matrixWithFreq <- merge(reqData,freqOfEachNode,by.x="fromReplacement",by.y="Var1")

#sparse Matrix representation
webGraph <- t(sparseMatrix(i = matrixWithFreq$fromReplacement,j = matrixWithFreq$toReplacement,x=1.0/matrixWithFreq$Freq))

#Number of nodes in this graph
noOfVertices <- nrow(webGraph)

#initializing page ranks with 1/noOfVetices
pageRanks <- rep(1/noOfVertices,noOfVertices)

#Iterative Calculation for Pageranks
for (i in 1:1000){
  tempPageRanks <- beta*as.numeric((webGraph %*% pageRanks)) + (1-beta)/noOfVertices
  rankSum = sum(tempPageRanks)
  tempPageRanks = tempPageRanks + (1-rankSum)/noOfVertices
  comparisonValue <- sum(abs(pageRanks - tempPageRanks))
  if(comparisonValue < 0.000001){
    pageRanks = tempPageRanks
    break
  }else{
    pageRanks = tempPageRanks
  }
}

#Finding replacement value for node 99
Node99Replacement = allNodesWithReplacement[allNodesWithReplacement$from==99]$fromReplacement[1]
#Printing page rank of Node 99
print(pageRanks[Node99Replacement])

File: MatrixFormulation/Python/UsingSparseMatrix.py
#importing required packages
import pandas as pd
import numpy as np
from scipy.sparse import csr_matrix

#Data from Google Page Rank Competition
webGoogle = pd.read_csv('../Graph/web-Google.txt', sep='\t', comment='#', names=['from', 'to'])

#Probability to visit a random link
beta = 0.8

#Link Number for each link in data
data = np.repeat(1, webGoogle.shape[0])
#Number of unique nodes in data
categories = np.unique(webGoogle.values)
#Sorting the categories for numbering them
categories.sort()
#Total Number of nodes in data
noOfNodes = len(categories)
#Code for each node in data
codes = np.arange(0,noOfNodes)
#Pandas Data Frame by replacing nodes with codes
replacementCodes = pd.DataFrame({"node" : categories, "code" : codes})
#Creating dictionary with categories as index and codes as values
replaceDict = dict(zip(categories,codes))
#Replacing FromNodes with Codes
fromReplaced = pd.merge(webGoogle,replacementCodes,left_on="from",right_on="node")
#Renaming for better understanding
fromReplaced.columns = ["from","to","fromReplacement","node"]
#Replacing toNodes with Codes
toFromReplaced = pd.merge(fromReplaced[["from","to","fromReplacement"]],replacementCodes,left_on="to",right_on="node")
#Renaming for better understanding
toFromReplaced.columns = ["from","to","fromReplacement","toReplacement","node"]
#From nodes are taken as rows
row_ind = toFromReplaced['fromReplacement']
#To nodes are taken as columns
col_ind = toFromReplaced['toReplacement']
#Frequency of each node
freq = np.bincount(row_ind)
nodes = np.nonzero(freq)[0]
freqWithNodes = pd.DataFrame({"node" : nodes,"frequency" : freq[nodes]})
#Merging Replaced Nodes with Codes and above frequency for each from node
preprocessedData = pd.merge(toFromReplaced,freqWithNodes,left_on="fromReplacement",right_on="node")

#Sparse Matrix with ToNodes as rows and FromNodes as columns
M = csr_matrix((1.0/preprocessedData['frequency'], (preprocessedData["fromReplacement"], preprocessedData["toReplacement"]))).transpose()

print M.shape
#Initializing pageRank for each Node with 1/noOfNodes
pageRanks = np.repeat(1/noOfNodes, noOfNodes)

#Iterative Calculation for Pageranks
for i in range(1000):
  tempPageRanks = beta*M.dot(pageRanks) + (1-beta)/noOfNodes
  rankSum = sum(tempPageRanks)
  tempPageRanks = tempPageRanks + (1-rankSum)/noOfNodes
  comparisonValue = sum(abs(pageRanks - tempPageRanks))
  if comparisonValue < 0.000001:
    pageRanks = tempPageRanks
    break
  else:
    pageRanks = tempPageRanks

#Page Rank of Specific Node with value 99 in the given dataset
print pageRanks[replaceDict.get(99)]



